### DISCUSSION TITLE
Allow Helm storage to be set to ConfigMap

### DISCUSSION THREAD

**Discussion by steenbjerg** [fluxcd/flux2#839] [Q&A] [answered]
hi - I would like to use helm-controller for deploying helm charts to our openshift environment. I have installed the installed flux-v2 and can deploy helm charts. Works fine. But we have some requirements, I need to take into account. The developers are not allowed to create k8s secrets and service-accounts (and some other kinds of objects). So the helm-controller must be prohibited in creating these kinds of objects. Therefore, I want change the service-account/roles that the helm-operator uses. This is possible, but unfortunately helm needs to create the version secrets (like sh.helm.release.v1.<application>.v1). When using Helm directly, it can use an environment variable named HELM_DRIVER, making helm use configmaps instead of secrets for the version history. But I cant set this environment variable at the helm-controller (if I set up an env at the helm-controller deployment, it has no effect on the helm behaviour). So I would like to know if this is possible?

**Comment by hiddeco** [comment, marked as answer]
This is not possible, nor are there plans to make this possible.

As a `HelmRelease` can refer to `Secret` resources, the composition of values may include data that is categorized as sensitive. Storing those in a `ConfigMap` would defeat the purpose of putting them in `Secret` resources, as they are not stored encrypted in etcd.

Can you provide a more detailed example of what your developers would enroll without having access to things like `Secret` and `ServiceAccount` resources?

**Reply by steenbjerg** [reply]
For instance developers must not know production passwords - only the security department. Both service accounts and secrets are created by special processes/applications with special permissions. Therefore, no helm charts include secrets/serviceaccounts.
That is our Helm releases do not refer to any sensitive data. I dont know the precise content of the helm version secret, but since our helm charts do not contain any sensitive data, my thought was to use a configmap instead.

**Comment by stefanprodan** [comment]
You can configure a service account under which helm-controller runs helm commands so you can define your own RBAC for each release, see https://toolkit.fluxcd.io/components/helm/helmreleases/#role-based-access-control

**Reply by steenbjerg** [reply]
My first idea were to add a serviceaccount/role/binding and make the helm-controller run as this service account. This role should prevent the helm-controller from creating secrets and/or serviceaccounts. But If I do that, the helm-controller will not be able to create the helm version secret. Can you see a way around this problem?

**Reply by stefanprodan** [reply]
I posted above the solution that we’re implemented for restricting the release operations with RBAC. 

TLDR: helm-controller impersonates a service account that you (the cluster admin) have set in the HelmRelease object. This has nothing to do with the service account of the helm-controller pod, you can define a custom RBAC per release.

**Reply by steenbjerg** [reply]
Thanks for your explaination. I didnot know that. I tried it. I created a service, role and rolebinding. The serviceaccount (named app-reconciler) is not allowed to create secrets. My chart does not include any secret yaml but still I get this log inside the helm-controller pod:

Helm install failed: create: failed to create: secrets is forbidden: User \"system:serviceaccount:playground-jb0278-2:app-reconciler\" cannot create resource \"secrets\" in API group \"\" in the namespace \"playground-jb0278-2\"

It still tries to create a secret - this must be the helm version secret. Do you have any suggestions how to avoid this? When using helm I have setup the HELM_DRIVER environment variable to force the helm version stuff to be written to a confgmap instead of a secret (my chart does not include any sensitive data).

**Reply by stefanprodan** [reply]
@steenbjerg can you please change the title to "Allow Helm storage to be set to ConfigMap" as the current one is miss-leading. You can block service accounts, roles and any other object besides secrets with the current impersonation implementation.

**Comment by ginn13** [comment]
@stefanprodan is there a way to use HELM_DRIVER  to use [SQL Storage Backend](https://helm.sh/docs/topics/advanced/#sql-storage-backend)? We reached the limit of 1 Megabyte size of Secret with helm release and searching the way to bypass this limit

**Reply by kennethredler** [reply]
Admittedly many solutions (charts, and how people choose to deploy helm releases) MAY manage Secrets including the sensitive information.

However, consider an implementation where External Secreta Operator is used to manage sensitive information into kubernetes Secrets from an external secret solution like AWS Secrets Manager, for example.

In this case only an ExternalSecret which DOES NOT contain sensitive information would be included in the Helm Release.

With this strategy, sensitive information is kept out of the helm release allowing it to be reviewed without the sensitivity and risk.

Do you see how in this case allowing the HelmRelease controller to use ConfigMap storage instead of Secrets would improve visibility without risk of exposing sensitive information in other Secrets  @stefanprodan  ?
